beacon_node/network/src/prefix_mapping.rs

```diff
@@ -18,66 +22,117 @@ impl PrefixMapping {
         &mut self,
         subnet_id: &SubnetId,
         current_epoch: Epoch,
-    ) -> Result<Vec<NodeId>, &'static str> {
+    ) -> Result<Vec<i32>, &'static str> {
         if self.epoch != current_epoch {
-            // compute prefixes
-            let computed_mapping = SubnetId::compute_prefix_mapping_for_epoch::<TSpec>(
+            // compute prefix mapping
+            self.mapping = SubnetId::compute_prefix_mapping_for_epoch::<TSpec>(
                 current_epoch,
                 &self.chain_spec,
             )?;
 
-            self.mapping = computed_mapping
-                .into_iter()
-                .map(|(subnet_id, ids)| {
-                    // convert `U256`s to `NodeId`s
-                    let node_ids = ids
-                        .into_iter()
-                        .map(|id| {
-                            let raw_node_id: [u8; 32] = id.into();
-                            NodeId::from(raw_node_id)
-                        })
-                        .collect::<Vec<_>>();
-
-                    (subnet_id, node_ids)
-                })
-                .collect();
-
             self.epoch = current_epoch;
         }
 
-        let node_ids = self
+        let prefixes = self
             .mapping
             .get(subnet_id)
-            .ok_or("No NodeId in the prefix mapping.")?
+            .ok_or("No prefix in the mapping.")?
             .clone();
 
-        Ok(node_ids)
+        Ok(prefixes)
+    }
+
+    /// Returns random `NodeId`s with the prefix that should be subscribed to the given `SubnetId` and `Epoch`.
+    pub fn get_target_nodes<TSpec: EthSpec>(
+        &mut self,
+        subnet_id: &SubnetId,
+        current_epoch: Epoch,
+    ) -> Result<Vec<NodeId>, &'static str> {
+        let subnet_prefix_bits = self.chain_spec.attestation_subnet_prefix_bits as u32
+            + self.chain_spec.attestation_subnet_shuffling_prefix_bits as u32;
+        // The `mask` is used in a bitwise operation to replace a specific segment of the NodeId
+        // with the subnet prefix bits.
+        // This variable indicates which segment of the NodeId should be replaced with the subnet
+        // prefix bits.
+        let mask = U256::from(2_i32.pow(subnet_prefix_bits) - 1) << (256 - subnet_prefix_bits);
+
+        Ok(self
+            .get_prefixes::<TSpec>(subnet_id, current_epoch)?
+            .into_iter()
+            .map(|prefix| {
+                // We generate a random NodeId and replace the first few bits with the prefix.
+                let random_node_id = U256::from(NodeId::random().raw());
+                let prefixed_node_id = U256::from(prefix) << (256 - subnet_prefix_bits);
+
+                // Replace the first few bits of the random NodeID with the prefix. The bits that we
+                // want to replace are identified by the `mask`.
+                let raw_node_id: [u8; 32] =
+                    (random_node_id ^ ((random_node_id ^ prefixed_node_id) & mask)).into();
```

Comment:
@AgeManning Thanks for your review! I have updated `PrefixMapping::get_target_nodes` to return random node-ids. https://github.com/sigp/lighthouse/pull/5090/commits/6f614df182c5d6237ce7fd3b2e481fd6aaff9def

The method used to replace the first few bits might seem a bit hacky. I'm not sure if this is the best approach, but I'm open to other suggestions.